src/libraries/System.Text.Json/src/System/Text/Json/Document/JsonDocument.cs

```diff
@@ -874,6 +874,55 @@ private static void ClearAndReturn(ArraySegment<byte> rented)
             }
         }
 
+        internal void WritePropertyName(int index, Utf8JsonWriter writer)
+        {
+            CheckNotDisposed();
+
+            WritePropertyName(_parsedData.Get(index - DbRow.Size), writer);
+        }
```

Comment:
hi @eiriktsarpalis 

`JsonProperty.WriteTo` will consume a thin wrapper (the method above) around `JsonDocument.WritePropertyName(in DbRow row, Utf8JsonWriter writer)`. 

However, I found `JsonDocument.WritePropertyName(in DbRow row, Utf8JsonWriter writer)` always rents an array regardless of length of property names. Though I observed it's usually allocation-less, due to reuse in `ArrayPool` I suppose.

So I wrote a stackalloc version (`WritePropertyNameNew` down below), which is:
* More verbose.
* Guaranteed to be allocation-less when the property name is short.
* ~10% faster when name contains escaped or Unicode characters (is it rare case?).
    * Speed is on par with the original version if there are no special chars.

I was deciding:
* Should we replace the original `JsonDocument.WritePropertyName`?
* I think similar optimizations can be done with `JsonDocument.WriteStringValue`. Should we replace that as well?

I wonder if there are already considerations on why `WritePropertyName` always utilizes rented array, and I would kindly ask for your opinions.

Reply:
I'm sorry to hear that and hope you get better soon. Rest assured there's no deadline involved here, feel free to pick this up again if and whenever you'd be inclined to continue it.

Reply:
Thank you. 

I just would like to inquire you, should we refactor the current `JsonDocument.WritePropertyName` to make use of `stackalloc` whenever possible?

* The current implementation always rents an array semantically.
* But the refactor will make the method much more verbose as the method `WritePropertyNameNew` I wrote in the draft code.